core/deployment/src/main/java/io/quarkus/deployment/steps/ApplicationIndexBuildStep.java

```diff
@@ -33,14 +40,33 @@ public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) th
                 @Override
                 public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
                     if (file.getFileName().toString().endsWith(".class")) {
-                        log.debugf("Indexing %s", file);
-                        try (InputStream stream = Files.newInputStream(file)) {
-                            indexer.index(stream);
+                        if (isRemovedApplicationClass(
+                                file.toString(), removedApplicationClasses)) {
+                            log.debugf("File %s will not be indexed because the class has been configured as part of '%s'",
+                                    file, "quarkus.class-loading.removed-resources");
+                        } else {
+                            log.debugf("Indexing %s", file);
+                            try (InputStream stream = Files.newInputStream(file)) {
+                                indexer.index(stream);
+                            }
                         }
                     }
                     return FileVisitResult.CONTINUE;
                 }
 
+                private boolean isRemovedApplicationClass(String fileName, Set<String> removedApplicationClasses) {
```

Comment:
Maybe path the original `Path` here and call `toString()` on it if `removedApplicationClasses` isn't empty?

Reply:
Sure, done